# Assignment 5

## 2015136097 이수현

---

---
### [문제 1]
#### 클래스와 모듈의 공통점과 차이점에 대해 설명하시오.

#### ::설명::
    모듈은 각각 구분이 가능한 자신만의 이름공간을 가져 이를 활용할 수 있는 다양한 방법들이 있습니다. 이 모듈 안에서도 모듈과 비슷한 역할을 하여 새로운 이름공간을 제공하는 단위가 클래스입니다. 클래스는 모듈 안의 클래스 영역으로 이름공간을 구성합니다. 클래스와 모듈 모두 이름공간을 제공한다는 점에서 공통점을 가집니다. 하지만 모듈은 파일 단위로 이름 공간을 구성하는 반면, 클래스는 클래스 영역과 인스턴스 영역 내에 이름 공간을 구성한다는 점에서 차이를 보입니다.

---
### [문제 2]
#### 다형성에 대해 설명하고 다형성을 보여주는 자신만의 파이썬 코드 예제를 제시하시오.

In [9]:
class Figure:
    def draw(self):
        print '도형 그리기'
        
class Rectangle(Figure):
    def draw(self):
        print '사각형 그리기'
        
class Triangle(Figure):
    def draw(self):
        print '삼각형 그리기'
        
class Circle(Figure):
    def draw(self):
        print '원 그리기'

for each in (Rectangle(), Triangle(), Circle()):
    each.draw()

사각형 그리기
삼각형 그리기
원 그리기


#### ::설명::
    다형성은 상속 관계 내의 다른 클래스들의 인스턴스들이 같은 멤버 함수 호출에 대해 각각 다르게 반응하도록 하는 기능입니다. 적은 코딩으로 다양한 객체들에게 유사한 작업을 수행시킬 수 있고, 프로그램 작성 코드 량이 줄어들고, 코드의 가독성을 높혀주는 장점을 가집니다. 파이썬에서는 형 선언이 없다는 점에서 파이썬에서는 다형성을 적용하기가 더욱 용이하고, 실시간으로 객체의 형이 결정되므로 단 하나의 메소드에 의해 처리될 수 있는 객체의 종류에 제한이 없다는 점에서 더욱 장점을 가집니다.

---
### [문제 3]
#### 다음 각 요구사항 모두를 만족시키는 Counter 클래스를 코딩하시오.(요구사항 생략)


In [7]:
class Counter : 
    def __init__(self, count, step=1):
        self.count = count
        self.step = step
    def __str__(self):
        return "[Count (step: "+ str(self.step) + ")] " + str(self.count)
    def incr(self):
        self.count = self.count+ self.step
    def __call__(self):
        self.incr()
    def __add__(self,num):
        return Counter(self.count+num,self.step)
    def __sub__(self,num):
        return Counter(self.count-num,self.step)   
    def __lt__(self, other):
        if self.count < other:
            return True
        else:
            return False
    def __gt__(self, other):
        if self.count > other:
            return True
        else:
            return False  
    def __eq__(self, other):
        if self.count == other:
            return True
        else:
            return False
    def __ne__(self, other):
        if self.count != other:
            return True
        else:
            return False
               

#### ::설명::
1. _ _init_ _을 사용하여 생성자를 정의하였습니다.
2. _ _str_ _을 사용하여 요구사항대로 출력할 수 있도록 하였습니다.
3. incr에서 count를 증가시킬 수 있도록 하였습니다.
4. _ _call_ _을 사용하여 인스턴스 객체를 직접 호출하면 내부적으로 self.incr()을 통해 count를 증가시키도록 하였습니다.
5. _ _add_ _ 등의 연산자 중복을 구현하였습니다.

---
### [문제 4]
#### 다음은 내장 자료형 list를 서브클래싱하여 만든 MySet 클래스 정의 내용이다. 다음 클래스 정의에서 __init__(), __str()__(), elimicate_duplicate()의 세 개의 메소드 코드 내용을 자신이 다른 사람에게 가르친다고 생각하며 설명해보시오.

- MySet은 집합(Set) 자료형을 정의하려는 의도하에 만들어진 클래스이다.
   

In [18]:
class MySet(list):
    def __init__(self, l):
        for e in l:
            self.append(e)
        MySet.eliminate_duplicate(self)
    
    def __str__(self):
        result = "MySet: {"
        for e in self:
            result = result + str(e) + " ,"
        result = result[0:len(result)-2] + "}"
        return result

    @staticmethod    
    def eliminate_duplicate(l):
        s = []
        for e in l:
            if e not in s:
                s.append(e)
        l[:] = []
        for e in s:
            l.append(e)

    
if __name__ == "__main__":
    s = MySet([1, 2, 2, 3])
    print s
    t = MySet([2, 3, 4, 5, 6, 7, 8, 8, 8, 8, 8, 9])
    print t

MySet: {1 ,2 ,3}
MySet: {2 ,3 ,4 ,5 ,6 ,7 ,8 ,9}


#### ::설명::
    init()은 클래스의 생성자입니다. s = MySet([1, 2, 2, 3])의 코드에서 [1, 2, 2, 3]을 받으면 자기 객체 s의 리스트에 1, 2, 2, 3을 추가한 후에 eliminate_duplicate(self)를 통해서 중복되는 원소를 제거합니다. 최종적으로 s는 [1,2,3]이 되는 것입니다. 이 때 eliminate_duplicate는 정적 메소드로 정의하였는데 정적 메소드란 인스턴스 객체와 무관하게 클래스 이름 공간에 존재하는 메소드로서 클래스 이름을 이용하여 직접 호출할 수 있는 메소드를 말합니다. eliminate_duplicate(I)를 보면 self가 없는 것을 볼 수 있습니다. str()은 print 예약어에 대응되는 메소드입니다. print s 를 했을때 s객체의 내용을 표시해주는 기능을 합니다. Myset은 집합 자료형을 정의하려는 메소드이기 때문에 str의 내용에서 집합 표시인 {} 형으로 표시 되도록 코딩한 것을 볼 수 있습니다.

---
### [문제 5]
#### 4번 문제에 정의된 MySet 클래스에 메소드를 추가하여 다음 각 요구사항 모두를 만족시키는 코딩을 제시하시오.


In [16]:
class MySet(list):
    def __init__(self, l):
        for e in l:
            self.append(e)
        MySet.eliminate_duplicate(self)
    
    def __str__(self):
        result = "MySet: {"
        for e in self:
            result = result + str(e) + " ,"
        result = result[0:len(result)-2] + "}"
        return result
    
    #요구사항1. 설명란에서 설명!
    def __or__(self, other):
        I = []
        for e in self:
            I.append(e)
        for e in other:
            I.append(e)
        MySet.eliminate_duplicate(I)
        return MySet(I)
    #요구사항2. 설명란에서 설명!
    def __and__(self, other):
        I = []
        for e in other:
            if e in self:
                I.append(e)
        return MySet(I)
    #요구사항3. 설명란에서 설명!    
    def __sub__(self, other):
        I = []
        for e in self:
            if e not in other:
                I.append(e)
        return MySet(I)

    @staticmethod    
    def eliminate_duplicate(l):
        s = []
        for e in l:
            if e not in s:
                s.append(e)
        l[:] = []
        for e in s:
            l.append(e)

    
if __name__ == "__main__":
    s = MySet([1, 2, 2, 3])
    print s
    t = MySet([2, 3, 4, 5, 6, 7, 8, 8, 8, 8, 8, 9])
    print t

MySet: {1 ,2 ,3}
MySet: {2 ,3 ,4 ,5 ,6 ,7 ,8 ,9}


#### ::설명::
1. [요구사항 1] | 연산자 중복정의를 위해 _ _or_ _ 을 사용하였습니다. self의 내용과 other의 내용을 전부 다 리스트 I에 넣은 후 eliminate_duplicate를 활용하여 중복되는 원소를 제거한 후 I를 이용한 새로운 객체를 반환하도록 코딩하였습니다.
2. [요구사항 2] & 연산자 중복정의를 위해 _ _and_ _ 을 사용하였습니다. other의 원소중에서 self에도 들어있는 원소가 있다면 I에 추가하여 교집합 I를 만든 후 I를 이용한 새로운 객체를 반환하도록 코딩하였습니다.
3. [요구사항 3] - 연산자 중복정의를 위해 _ _sub_ _ 을 사용하였습니다. self의 원소 중에 other에 포함되지 않는 원소들만 I에 추가하여 차집합 I를 만든 후 I를 이용한 새로운 객체를 반환하도록 코딩하였습니다.

---
### [문제 6]
#### 5번 문제에서 정의한 MySet 클래스에 대해 다음 예제를 수행하면 오류없이 올바르게 동작하는 것을 확인할 수 있다. 다음 예제 내에 있는 len(), bool() 내장함수와 in 키워드 사용 예제가 별다른 메소드 정의를 하지 않았는 데도 올바르게 수행되는 이유를 설명하시오.

In [17]:
s = MySet([1, 2, 3, 4, 5, 6])

print len(s)

print bool(s)

print 2 in s


6
True
True


#### ::설명::
    len, bool, in 등이 내장함수이기 때문입니다. 내장 함수란 파이썬의 주요 함수로써, 이름 공간에 명시적으로 들여올 필요 없이 언제든지 사용할 수 있는 것들이기 때문에 별다른 메소드 정의를 하지 않아도 올바르게 코드가 수행됩니다.

---
### [프로젝트 오일러 16]
#### 215 = 32768 의 각 자리수를 더하면 3 + 2 + 7 + 6 + 8 = 26 입니다.

#### 21000의 각 자리수를 모두 더하면 얼마입니까?


In [1]:
s = str(2**1000)
L =[]
for i in range(0, len(s)):
    L.append(int(s[i]))
    
print sum(L)


1366


#### ::설명::
    2^1000의 값을 문자열 형태로 s에 저장한 후 인덱스를 하나씩 늘려가며 0부터 마지막까지 숫자 하나하나를 다시 int형으로 변환하여 리스트에 추가해주었습니다. 그 후 sum을 통해 리스트의 모든 값을 더하여 출력하면 각 자리수를 모두 더한 값이 출력됩니다.
#### 답: 1366

---
### [프로젝트 오일러 17]
#### 1부터 5까지의 숫자를 영어로 쓰면 one, two, three, four, five 이고,
#### 각 단어의 길이를 더하면 3 + 3 + 5 + 4 + 4 = 19 이므로 사용된 글자는 모두 19개입니다.

#### 1부터 1,000까지 영어로 썼을 때는 모두 몇 개의 글자를 사용해야 할까요?


In [3]:
D = {0:0,1:3,2:3,3:5,4:4,5:4,6:3,7:5,8:5,9:4,10:3,11:6,12:6,13:8,14:8,15:7,16:7,17:9,18:8,19:8,20:6,30:6,40:5,50:5,60:5,70:7,80:6,90:6,100:7,1000:8}

mynum = 0

for i in range(1, 1001):
        if i <= 20 :
            mynum = mynum + D[i]
        elif i>20 and i<100 :
            mynum = mynum + D[(i/10)*10] + D[i%10]
        elif i>=100 and i<1000 :
            if i%100 != 0 :
                if i%100 > 20 :
                    mynum = mynum + D[(i/100)] + D[100] + 3 + D[(i%100)/10*10] + D[i%10]
                elif i%100 <= 20 :
                    mynum = mynum + D[(i/100)] + D[100] + 3 + D[i%100]
            else :
                mynum = mynum + D[(i/100)] + D[100]
        else :
            mynum = mynum + D[(i/1000)] + D[1000]
        
print mynum

21124


#### ::설명::
1. 먼저 조건에 맞게 적절히 숫자를 세기 위해 1부터 1000까지 중 추가해야 할 숫자들만 사전에 숫자와 영어로 썼을때의 글자 수를 묶어 추가하였습니다.
2. i를 1부터 1000까지 반복하여 숫자에 따라서 사전을 각각 다르게 조합해 단어 수를 세었습니다.
3. 20 이하일때는 사전에 넣어진 단어 수 그대로를 mynum(총 단어 수)에 추가합니다. 20초과 100 미만일때는 10의자리수 + 1의자리수로 추가합니다. 100이상 1000미만일때는 100으로 나누어 떨어지는것과 아닌것을 나누어서 더했습니다. 1000일때는 1000의 단어 수를 추가하여 총 단어 수를 구하였습니다.
#### 답 : 21124

---
### [프로젝트 오일러 18]
#### 삼각형의 꼭대기부터 아래쪽으로 인접한 숫자를 찾아 내려가면서 합을 구하면, 위의 그림처럼 3 + 7 + 4 + 9 = 23 이 가장 큰 합을 갖는 경로가 됩니다. 

#### 다음 삼각형에서 합이 최대가 되는 경로를 찾아서 그 합을 구하세요. (삼각형 생략)

In [4]:
I = [[75], [95, 64], [17, 47, 82], [18, 35, 87, 10], [20, 4, 82, 47, 65], [19, 1, 23, 75, 3, 34], [88, 2, 77, 73, 7, 63, 67], [99, 65, 4, 28, 6, 16, 70, 92], [41, 41, 26, 56, 83, 40, 80, 70, 33], [41, 48, 72, 33, 47, 32, 37, 16, 94, 29], [53, 71, 44, 65, 25, 43, 91, 52, 97, 51, 14], [70, 11, 33, 28, 77, 73, 17, 78, 39, 68, 17, 57], [91, 71, 52, 38, 17, 14, 91, 43, 58, 50, 27, 29, 48], [63, 66, 4, 68, 89, 53, 67, 30, 73, 16, 69, 87, 40, 31], [4, 62, 98, 27, 23, 9, 70, 98, 73, 93, 38, 53, 60, 4, 23]]

L = []

for n1 in range(0, 1) :
    for n2 in range(n1, n1+2) :
        for n3 in range(n2, n2+2) :
            for n4 in range(n3, n3+2) :
                for n5 in range(n4, n4+2) :
                    for n6 in range(n5, n5+2) :
                        for n7 in range(n6, n6+2) :
                            for n8 in range(n7, n7+2) :
                                for n9 in range(n8, n8+2) :
                                    for n10 in range(n9, n9+2) :
                                        for n11 in range(n10, n10+2) :
                                            for n12 in range(n11, n11+2) :
                                                for n13 in range(n12, n12+2) :
                                                    for n14 in range(n13, n13+2) :
                                                        for n15 in range(n14, n14+2) :
                                                            L.append(I[0][n1]+I[1][n2]+I[2][n3]+I[3][n4]+I[4][n5]+I[5][n6]+I[6][n7]+I[7][n8]+I[8][n9]+I[9][n10]+I[10][n11]+I[11][n12]+I[12][n13]+I[13][n14]+I[14][n15])
                 
                 

L.sort(reverse=True)

print L[0]
                 

1074


#### ::설명::
1. 삼각형을 줄 단위로 리스트에 저장합니다.
2. 15중 반복문을 사용하여 자신의 윗 줄의 해당 숫자의 인덱스, 해당 숫자의 인덱스+1 에만 접근할 수 있도록 범위를 설정하고 15줄을 한 숫자씩 더하였습니다.
3. 더한 숫자는 L에 저장하고, L을 내림차순 정렬하여 L[0]을 출력하면 가장 큰 숫자가 출력됩니다.
#### 답 : 1074

---
### [문제 10]
#### 이전 Assignment 3 (Assignment 4가 아님)의 마지막 문제는 웹 URL로 지정된 웹페이지를 문자열로 가져와 모든 HTML 태그 및 CSS와 Javascript를 제외한 순수 텍스트를 얻어내고 그 안에 존재하는 단어를 추출하여 각 단어들에 대해 출현빈도를 사전형태({'world': 2, 'hello': 1, 'python': 1})로 저장하여 출력하는 것이었다. 이번에는 Assignment 3을 다시 확장/변형하여 다음과 같은 조건을 만족하도록 구현하시오.


In [3]:
import urllib2
import string

#source = [urllib2.urlopen("http://www.ikonet.com/en/visualdictionary/").read(), \
#          urllib2.urlopen("http://theoatmeal.com/tag/grammar").read(), \
#          urllib2.urlopen("http://www.bloomberg.com/live/asia").read(), \
#          urllib2.urlopen("http://www.esl-lab.com/").read(), \
#          urllib2.urlopen("http://www.usalearns.org/").read()]

def cmp_1(a1, a2):
    return cmp(a2[1], a1[1])

class WebWordsFrequency(list):
    def __init__(self, *url):
        for i in range(0,len(url)):
            self.append(url[i])
    def addUrl(self, url):
        self.append(url)
    def removeUrl(self, url):
        self.remove(url)
    def listUrls(self):
        for i in self:
            print i
    def getWordsFrequency(self):
        stop = ['a', 'about', 'above', 'across', 'after', 'again', 'against', 'all', 'almost', 'alone', 'along', 'already', 'also', 'although', 'always', 'among', 'an', 'and', 'another', 'any', 'anybody', 'anyone', 'anything', 'anywhere', 'are', 'area', 'areas', 'around', 'as', 'ask', 'asked', 'asking', 'asks', 'at', 'away', 'b', 'back', 'backed', 'backing', 'backs', 'be', 'became', 'because', 'become', 'becomes', 'been', 'before', 'began', 'behind', 'being', 'beings', 'best', 'better', 'between', 'big', 'both', 'but', 'by', 'c', 'came', 'can', 'cannot', 'case', 'cases', 'certain', 'certainly', 'clear', 'clearly', 'come', 'could', 'd', 'did', 'differ', 'different', 'differently', 'do', 'does', 'done', 'down', 'down', 'downed', 'downing', 'downs', 'during', 'e', 'each', 'early', 'either', 'end', 'ended', 'ending', 'ends', 'enough', 'even', 'evenly', 'ever', 'every', 'everybody', 'everyone', 'everything', 'everywhere', 'f', 'face', 'faces', 'fact', 'facts', 'far', 'felt', 'few', 'find', 'finds', 'first', 'for', 'four', 'from', 'full', 'fully', 'further', 'furthered', 'furthering', 'furthers', 'g', 'gave', 'general', 'generally', 'get', 'gets', 'give', 'given', 'gives', 'go', 'going', 'good', 'goods', 'got', 'great', 'greater', 'greatest', 'group', 'grouped', 'grouping', 'groups', 'h', 'had', 'has', 'have', 'having', 'he', 'her', 'here', 'herself', 'high', 'high', 'high', 'higher', 'highest', 'him', 'himself', 'his', 'how', 'however', 'i', 'if', 'important', 'in', 'interest', 'interested', 'interesting', 'interests', 'into', 'is', 'it', 'its', 'itself', 'j', 'just', 'k', 'keep', 'keeps', 'kind', 'knew', 'know', 'known', 'knows', 'l', 'large', 'largely', 'last', 'later', 'latest', 'least', 'less', 'let', 'lets', 'like', 'likely', 'long', 'longer', 'longest', 'm', 'made', 'make', 'making', 'man', 'many', 'may', 'me', 'member', 'members', 'men', 'might', 'more', 'most', 'mostly', 'mr', 'mrs', 'much', 'must', 'my', 'myself', 'n', 'necessary', 'need', 'needed', 'needing', 'needs', 'never', 'new', 'new', 'newer', 'newest', 'next', 'no', 'nobody', 'non', 'noone', 'not', 'nothing', 'now', 'nowhere', 'number', 'numbers', 'o', 'of', 'off', 'often', 'old', 'older', 'oldest', 'on', 'once', 'one', 'only', 'open', 'opened', 'opening', 'opens', 'or', 'order', 'ordered', 'ordering', 'orders', 'other', 'others', 'our', 'out', 'over', 'p', 'part', 'parted', 'parting', 'parts', 'per', 'perhaps', 'place', 'places', 'point', 'pointed', 'pointing', 'points', 'possible', 'present', 'presented', 'presenting', 'presents', 'problem', 'problems', 'put', 'puts', 'q', 'quite', 'r', 'rather', 'really', 'right', 'right', 'room', 'rooms', 's', 'said', 'same', 'saw', 'say', 'says', 'second', 'seconds', 'see', 'seem', 'seemed', 'seeming', 'seems', 'sees', 'several', 'shall', 'she', 'should', 'show', 'showed', 'showing', 'shows', 'side', 'sides', 'since', 'small', 'smaller', 'smallest', 'so', 'some', 'somebody', 'someone', 'something', 'somewhere', 'state', 'states', 'still', 'still', 'such', 'sure', 't', 'take', 'taken', 'than', 'that', 'the', 'their', 'them', 'then', 'there', 'therefore', 'these', 'they', 'thing', 'things', 'think', 'thinks', 'this', 'those', 'though', 'thought', 'thoughts', 'three', 'through', 'thus', 'to', 'today', 'together', 'too', 'took', 'toward', 'turn', 'turned', 'turning', 'turns', 'two', 'u', 'under', 'until', 'up', 'upon', 'us', 'use', 'used', 'uses', 'v', 'very', 'w', 'want', 'wanted', 'wanting', 'wants', 'was', 'way', 'ways', 'we', 'well', 'wells', 'went', 'were', 'what', 'when', 'where', 'whether', 'which', 'while', 'who', 'whole', 'whose', 'why', 'will', 'with', 'within', 'without', 'work', 'worked', 'working', 'works', 'would', 'x', 'y', 'year', 'years', 'yet', 'you', 'young', 'younger', 'youngest', 'your', 'yours', 'z']
        t = string.punctuation
        I = [e for e in t]
        L = []
        source = [urllib2.urlopen(i).read() for i in self]
        for i in range(0, len(source)):
            x = 0
            s = ""
            while( x != len(source[i])):
                if source[i][x] == '<':
                    if source[i][x+1:x+7] == 'script':
                        while source[i][x-8:x+1] != '</script>':
                            x += 1
                        if source[i][x] == '>' and x != len(source[i]):
                            x += 1
                    elif source[i][x+1:x+6] == 'style':
                        while source[i][x-7:x+1] != '</style>':
                            x += 1
                        if source[i][x] == '>' and x != len(source[i]):
                            x += 1
                    else:
                        while source[i][x] != '>':
                            x += 1
                        if source[i][x] == '>' and x != len(source[i]):
                            x += 1
                elif source[i][x:x+4] == '&amp':
                    x += 4
                else :
                    if source[i][x] in I:
                        x += 1
                    else:
                        s = s+source[i][x]
                        x += 1
    
            L.append(s.split())
            
    
        #영어 불용어들 제거하는 코드
        L2 = []
        for x in L:
            for y in x:
                Y = y.lower()
                if not Y in stop:
                    L2.append(y)
    
        #공백 단위로 단어 쪼개고 단어 넣을 딕셔너리 선언
        words = {}
 
        #단어 쪼개놓은 리스트에서 각각 단어의 개수 세기
        for i in L2 :
            if not i in words.keys():
                words[i]=1
            else:
                words[i]=words[i]+1

        return words
    
    def getMaxFreqencyWords(self):
        L = (self.getWordsFrequency()).items()
        if not L:
            return None #비어있으면 None
        L.sort(cmp_1)
        isTop = True
        i = 0
        L2 = []
        while isTop :
            #같은 개수면 같은 개수인것 다 출력
            L2.append(L[i][0])
            if L[i][1] != L[i+1][1] :
                isTop = False
            i += 1
        return L2
    
    def searchUrlByWord(self, word):
        List = []
        L2 = []
        for i in range(0, len(self)):
            List.append(self[i])
        for i in List :
            self.remove(i)
        for i in range(0, len(List)):
            self.append(List[i])
            L = (self.getWordsFrequency()).items()
            if not L:
                return None #비어있으면 None
            L.sort(cmp_1)
            L2.append(L)
            self.remove(List[i])
        for i in List :
            self.append(i)
        Sim = {}
        for i in self:
            Sim[i] = 0
        for x in range(0,len(L2)):
            for i in range(0,len(L2[x])):
                if word == L2[x][i][0]:
                    Sim[self[x]] = L2[x][i][1]
        L3 = Sim.items()
        L3.sort(cmp_1)
        isTop = True
        i = 0
        MyList = []
        while isTop :
            #같은 유사도이면 다 출력
            MyList.append(L3[i][0])
            if i == len(L3)-1 or L3[i][1] != L3[i+1][1] :
                isTop = False
            i += 1
        return MyList
        
w1 = WebWordsFrequency('http://www.cnn.com','http://www.ikonet.com/en/visualdictionary/','http://theoatmeal.com/tag/grammar') 
w2 = WebWordsFrequency('http://www.cnn.com', 'http://www.ikonet.com/en/visualdictionary/') 
w3 = WebWordsFrequency()

w1.addUrl('http://www.esl-lab.com')

w1.listUrls()

print w1.getWordsFrequency()

print w1.getMaxFreqencyWords()
print w3.getMaxFreqencyWords()
print w1.searchUrlByWord('Web')

http://www.cnn.com
http://www.ikonet.com/en/visualdictionary/
http://theoatmeal.com/tag/grammar
http://www.esl-lab.com
{'Friday': 1, 'Snack': 1, 'corps': 1, 'httpwwwquestionprocomaTakeSurvey': 1, 'Boutique': 1, 'Texting': 1, 'themes': 1, 'jeunesse': 1, 'Trivia': 1, 'Business': 3, 'Kicking': 1, 'Object': 1, 'Trees': 1, '0': 2, 'Bookstore': 1, 'Theft': 1, 'explode': 1, 'Funerals': 1, 'Day': 5, 'activities': 1, 'Grocery': 1, 'NationVital': 2, 'Condolences': 1, 'Sky': 1, 'RANDOM': 1, 'Internationalnbsp': 1, 'Candy': 1, 'Conversation': 1, 'Tech': 1, 'artwork': 1, 'Economy': 1, 'Restaurants': 1, 'Exotic': 1, 'illustrated': 1, 'Hamburger': 1, 'Crowd': 1, 'Hiking': 3, 'AllnbspRightsnbspReservedCNNnbspSansnbsptradenbspnbspcopynbsp2016nbspCablenbspNewsnbspNetworkTerms': 1, 'skills': 1, 'rights': 1, 'Trail': 1, 'Pies': 1, 'configskipCount': 1, 'ABCs': 1, 'Text': 1, 'study': 1, 'Breakfast': 1, 'Systems': 2, 'Town': 4, 'CODE': 1, 'Spending': 1, 'Interviews': 1, 'Accent': 4, 'School': 5, 'servicenbs

#### ::설명::
    클래스를 list클래스를 상속하도록 코딩하였습니다. 문제에서 요구한 여러 메소드를 정의하였습니다. 삭제나 삽입 메소드는 리스트 내부함수로 구현할 수 있었습니다. 단어 분석 메소드들은 이전 과제의 코드에서 종합적으로 묶어서 하도록 수정하고 그 외의 코드는 동일하도록 구성하였습니다. 유사도는 찾는 단어와 일치하는지 유사도를 키 값을 사이트로한 유사도를 정리한 사전을 만든 후 유사도가 높은 사이트를 출력하도록 하였습니다.

---
### [문제 11]
#### 위 10번 문제에서 정의한 WebWordsFrequency 클래스를 상속하여 OrderedWebWordsFrequency 클래스를 정의하고 슈퍼클래스에 정의된 getWordsFrequency() 메소드를 오버라이드 하여 단어 출현 빈도를 내림 차순으로 정렬하여 리스트로 출력하시오.



In [5]:
def cmp_2(a1, a2):
    return cmp(a1[1], a2[1])
 
class OrderedWebWordsFrequency(WebWordsFrequency):
    def getWordsFrequency(self,reverse=False):
        stop = ['a', 'about', 'above', 'across', 'after', 'again', 'against', 'all', 'almost', 'alone', 'along', 'already', 'also', 'although', 'always', 'among', 'an', 'and', 'another', 'any', 'anybody', 'anyone', 'anything', 'anywhere', 'are', 'area', 'areas', 'around', 'as', 'ask', 'asked', 'asking', 'asks', 'at', 'away', 'b', 'back', 'backed', 'backing', 'backs', 'be', 'became', 'because', 'become', 'becomes', 'been', 'before', 'began', 'behind', 'being', 'beings', 'best', 'better', 'between', 'big', 'both', 'but', 'by', 'c', 'came', 'can', 'cannot', 'case', 'cases', 'certain', 'certainly', 'clear', 'clearly', 'come', 'could', 'd', 'did', 'differ', 'different', 'differently', 'do', 'does', 'done', 'down', 'down', 'downed', 'downing', 'downs', 'during', 'e', 'each', 'early', 'either', 'end', 'ended', 'ending', 'ends', 'enough', 'even', 'evenly', 'ever', 'every', 'everybody', 'everyone', 'everything', 'everywhere', 'f', 'face', 'faces', 'fact', 'facts', 'far', 'felt', 'few', 'find', 'finds', 'first', 'for', 'four', 'from', 'full', 'fully', 'further', 'furthered', 'furthering', 'furthers', 'g', 'gave', 'general', 'generally', 'get', 'gets', 'give', 'given', 'gives', 'go', 'going', 'good', 'goods', 'got', 'great', 'greater', 'greatest', 'group', 'grouped', 'grouping', 'groups', 'h', 'had', 'has', 'have', 'having', 'he', 'her', 'here', 'herself', 'high', 'high', 'high', 'higher', 'highest', 'him', 'himself', 'his', 'how', 'however', 'i', 'if', 'important', 'in', 'interest', 'interested', 'interesting', 'interests', 'into', 'is', 'it', 'its', 'itself', 'j', 'just', 'k', 'keep', 'keeps', 'kind', 'knew', 'know', 'known', 'knows', 'l', 'large', 'largely', 'last', 'later', 'latest', 'least', 'less', 'let', 'lets', 'like', 'likely', 'long', 'longer', 'longest', 'm', 'made', 'make', 'making', 'man', 'many', 'may', 'me', 'member', 'members', 'men', 'might', 'more', 'most', 'mostly', 'mr', 'mrs', 'much', 'must', 'my', 'myself', 'n', 'necessary', 'need', 'needed', 'needing', 'needs', 'never', 'new', 'new', 'newer', 'newest', 'next', 'no', 'nobody', 'non', 'noone', 'not', 'nothing', 'now', 'nowhere', 'number', 'numbers', 'o', 'of', 'off', 'often', 'old', 'older', 'oldest', 'on', 'once', 'one', 'only', 'open', 'opened', 'opening', 'opens', 'or', 'order', 'ordered', 'ordering', 'orders', 'other', 'others', 'our', 'out', 'over', 'p', 'part', 'parted', 'parting', 'parts', 'per', 'perhaps', 'place', 'places', 'point', 'pointed', 'pointing', 'points', 'possible', 'present', 'presented', 'presenting', 'presents', 'problem', 'problems', 'put', 'puts', 'q', 'quite', 'r', 'rather', 'really', 'right', 'right', 'room', 'rooms', 's', 'said', 'same', 'saw', 'say', 'says', 'second', 'seconds', 'see', 'seem', 'seemed', 'seeming', 'seems', 'sees', 'several', 'shall', 'she', 'should', 'show', 'showed', 'showing', 'shows', 'side', 'sides', 'since', 'small', 'smaller', 'smallest', 'so', 'some', 'somebody', 'someone', 'something', 'somewhere', 'state', 'states', 'still', 'still', 'such', 'sure', 't', 'take', 'taken', 'than', 'that', 'the', 'their', 'them', 'then', 'there', 'therefore', 'these', 'they', 'thing', 'things', 'think', 'thinks', 'this', 'those', 'though', 'thought', 'thoughts', 'three', 'through', 'thus', 'to', 'today', 'together', 'too', 'took', 'toward', 'turn', 'turned', 'turning', 'turns', 'two', 'u', 'under', 'until', 'up', 'upon', 'us', 'use', 'used', 'uses', 'v', 'very', 'w', 'want', 'wanted', 'wanting', 'wants', 'was', 'way', 'ways', 'we', 'well', 'wells', 'went', 'were', 'what', 'when', 'where', 'whether', 'which', 'while', 'who', 'whole', 'whose', 'why', 'will', 'with', 'within', 'without', 'work', 'worked', 'working', 'works', 'would', 'x', 'y', 'year', 'years', 'yet', 'you', 'young', 'younger', 'youngest', 'your', 'yours', 'z']
        t = string.punctuation
        I = [e for e in t]
        L = []
        source = [urllib2.urlopen(i).read() for i in self]
        for i in range(0, len(source)):
            x = 0
            s = ""
            while( x != len(source[i])):
                if source[i][x] == '<':
                    if source[i][x+1:x+7] == 'script':
                        while source[i][x-8:x+1] != '</script>':
                            x += 1
                        if source[i][x] == '>' and x != len(source[i]):
                            x += 1
                    elif source[i][x+1:x+6] == 'style':
                        while source[i][x-7:x+1] != '</style>':
                            x += 1
                        if source[i][x] == '>' and x != len(source[i]):
                            x += 1
                    else:
                        while source[i][x] != '>':
                            x += 1
                        if source[i][x] == '>' and x != len(source[i]):
                            x += 1
                elif source[i][x:x+4] == '&amp':
                    x += 4
                else :
                    if source[i][x] in I:
                        x += 1
                    else:
                        s = s+source[i][x]
                        x += 1
    
            L.append(s.split())
            
    
        #영어 불용어들 제거하는 코드
        L2 = []
        for x in L:
            for y in x:
                Y = y.lower()
                if not Y in stop:
                    L2.append(y)
                    
        #공백 단위로 단어 쪼개고 단어 넣을 딕셔너리 선언
        words = {}
 
        #단어 쪼개놓은 리스트에서 각각 단어의 개수 세기
        for i in L2 :
            if not i in words.keys():
                words[i]=1
            else:
                words[i]=words[i]+1
 
        MyList = words.items()
        if reverse==True:
            MyList.sort(cmp_2)
        else:
            MyList.sort(cmp_1)
        
        return MyList
    
    #def __iter__(self):
        #self.index=0
        #mylist = self.getWordsFrequency()
        #return self
    #def next(self):
        #if self.index>=len(mylist):
            #raise StopIteration
        #i = mylist[self.index]
        #self.index += 1
        #return i      
    
w4 = OrderedWebWordsFrequency('http://www.cnn.com','http://www.ikonet.com/en/visualdictionary/','http://theoatmeal.com/tag/grammar')
print w4.getWordsFrequency()
print w4.getWordsFrequency(reverse=True)
#for i in w4:
       #print i

[('Visual', 8), ('View', 8), ('Comics', 5), ('Online', 4), ('edition', 4), ('Dictionary', 4), ('Le', 4), ('Grammar', 3), ('nbsp', 3), ('Start', 3), ('Visuel', 3), ('CNN', 3), ('0', 2), ('Livres', 2), ('TeamBusinessMarketsTechLuxuryEntertainmentCelebrity', 2), ('degreesAll', 2), ('EditionUSInternationalArabicEspa\xc3\xb1olSet', 2), ('FeaturesVideoVideo', 2), ('WorldInside', 2), ('2016', 2), ('kingdom', 2), ('Shop', 2), ('ProjectImpact', 2), ('iPad', 2), ('SecurityInnovation', 2), ('NationVital', 2), ('Body', 2), ('English', 2), ('WatchTV', 2), ('newsFeature', 2), ('guide', 2), ('Contact', 2), ('Human', 2), ('ResultsNationWorldOur', 2), ('image', 2), ('ProfilesLeadershipPressroomPartner', 2), ('EastOpinionUS', 2), ('WorldwideMore\xe2\x80\xa6WeatherCNN', 2), ('comic', 2), ('TravellerDestinationsFeaturesFoodDrinkHotelsPartner', 2), ('store', 2), ('showsTV', 2), ('MusicGadgetsCyber', 2), ('Africa2', 2), ('Language', 2), ('WebMovies', 2), ('copy', 2), ('SignsSportFootballGolfTennisMotorsport

#### ::설명::
    단어 출현 빈도를 내림차순으로 정렬하여 리스트로 출력하는 것은 super 클래스의 코드를 그대로 따와 cmp_2 만 새로 추가하여 사전이 아닌 리스트로 출력하도록 수정해주었습니다. reverse 인자값은 if구문을 활용하여 코딩하였습니다.

---
### [문제 12]
#### 다음과 같은 코딩이 가능하도록 OrderedWebWordsFrequency 안에 반복자와 관련된 메소드를 추가하시오.
    >>> for i in w4:
    >>>       print i
    ('site, 12)
    ('science', 11)
    ('hello', 8)
    ('world', 2)
    ('program', 1)
    ('python', 1)

#### ::설명::
    __ iter __ 와 __ next __를 추가하여 구현하였습니다. 클래스 내부 다른 함수에 self를 사용하는 for in 구문이 너무 많아서 문제 11에 주석으로 추가해두었습니다.

---
### [느낀점]

#### 마지막 과제를 완성하니 마음이 뿌듯합니다! 한 학기동안 모두 수고 많으셨습니다!!
---